arangod/RestHandler/RestDumpHandler.cpp

```diff
@@ -0,0 +1,288 @@
+  }
+
+  if (ServerState::instance()->isCoordinator()) {
+    ServerID const& DBserver = _request->value("dbserver");
+    if (!DBserver.empty()) {
+      // if DBserver property present, remove auth header
+      _request->addHeader(StaticStrings::DumpAuthUser, _request->user());
+      return std::make_pair(DBserver, true);
+    }
+
+    return ResultT<std::pair<std::string, bool>>::error(
+        TRI_ERROR_BAD_PARAMETER, "need a 'dbserver' parameter");
+  }
+
+  return {std::make_pair(StaticStrings::Empty, false)};
+}
+
+void RestDumpHandler::handleCommandDumpStart() {
+  bool parseSuccess = false;
+  VPackSlice body = this->parseVPackBody(parseSuccess);
+  if (!parseSuccess) {  // error message generated in parseVPackBody
+    return;
+  }
+
+  auto database = _request->databaseName();
+  auto user = getAuthorizedUser();
+
+  RocksDBDumpContextOptions opts;
+  velocypack::deserializeUnsafe(body, opts);
+
+  auto& engine =
+      server().getFeature<EngineSelectorFeature>().engine<RocksDBEngine>();
+  auto* manager = engine.dumpManager();
+  auto guard = manager->createContext(std::move(opts), user, database);
+
+  resetResponse(rest::ResponseCode::CREATED);
+  _response->setHeaderNC(StaticStrings::DumpId, guard->id());
+}
+
+void RestDumpHandler::handleCommandDumpNext() {
+  TRI_ASSERT(!ServerState::instance()->isCoordinator());
+
+  auto const& suffixes = _request->suffixes();
+  // checked before
+  TRI_ASSERT(suffixes.size() == 2);
+  auto const& id = _request->suffixes()[1];
+
+  auto database = _request->databaseName();
+  auto user = getAuthorizedUser();
+
+  auto batchId = _request->parsedValue<uint64_t>("batchId");
+  if (!batchId.has_value()) {
+    generateError(Result(TRI_ERROR_BAD_PARAMETER, "expecting 'batchId'"));
+    return;
+  }
+
+  auto lastBatch = _request->parsedValue<uint64_t>("lastBatch");
+
+  auto& engine =
+      server().getFeature<EngineSelectorFeature>().engine<RocksDBEngine>();
+  auto* manager = engine.dumpManager();
+  // find() will throw in case the context cannot be found or the user does not
+  // match.
+  auto context = manager->find(id, database, user);
+  context->extendLifetime();
+
+  auto batch = context->next(*batchId, lastBatch);
+  auto counts = context->getBlockCounts();
+
+  if (batch == nullptr) {
+    // all batches have been received
+    return generateOk(rest::ResponseCode::NO_CONTENT, VPackSlice::noneSlice());
+  }
+
+  // output the batch value
+  _response->setHeaderNC(StaticStrings::DumpShardId, std::string{batch->shard});
+  _response->setHeaderNC(StaticStrings::DumpBlockCounts,
+                         std::to_string(counts));
+  _response->setContentType(rest::ContentType::DUMP);
+  _response->addRawPayload(batch->content);
+  _response->setGenerateBody(true);
+  _response->setResponseCode(rest::ResponseCode::OK);
+
+  // prolong lifetime of context, so that it is still there for follow-up
+  // requests.
+  context->extendLifetime();
```

Comment:
We already have a call to `extendLifetime()` in line 185. Is this intentional that it is called again?

Reply:
It actually was intentional, but @maierlars seems to have removed it now...
The rationale was to extend the lifetime as soon as the context is leased from the registry, so it is not garbage-collected while it is in use.
And then to extend its lifetime once again when it is "returned" to the registry, because the time a context is used can be pretty long (in case of slow IO or server overload).

Reply:
yeah, I was thinking about that. I was unsure if _that_ was the reason or if it was just an oversight. Feel free to add it again.